src/os/macos/lwpa/os_netint.c

```diff
@@ -0,0 +1,715 @@
+}
+
+/* Get the route destination from a normal socket address structure packed as part of a route
+ * entry.
+ */
+static void dest_from_route_entry(int family, const struct sockaddr* os_dst, LwpaIpAddr* lwpa_dst)
+{
+  if (family == AF_INET6)
+    LWPA_IP_SET_V6_ADDRESS(lwpa_dst, ((struct sockaddr_in6*)os_dst)->sin6_addr.s6_addr);
+  else
+    LWPA_IP_SET_V4_ADDRESS(lwpa_dst, ntohl(((struct sockaddr_in*)os_dst)->sin_addr.s_addr));
+}
+
+/* Get the netmask from a socket address structure packed as part of a route entry.
+ *
+ * The netmask sockaddrs take the form of a struct sockaddr, but they are incomplete (they are
+ * variable-length and not always the full length of the struct). The length field of the sockaddr
+ * correctly indicates where the structure is truncated, EXCEPT for the special case of a zero
+ * length, which indicates that the structure contains no other information (an implied zero
+ * netmask) but nevertheless occupies four bytes.
+ *
+ * If you have made it this far without throwing a full wine bottle at your computer screen, I
+ * congratulate you.
+ *
+ * This method for determining an IPv4 netmask comes from the Unix Network Programming book, sect.
+ * 18.3. It also references TCP/IP Illustrated Volume 2, fig. 18.21, which presents the way this
+ * data is held in the kernel. The method for obtaining an IPv6 netmask is undocumented anywhere,
+ * but inferred from the IPv4 method and empirical debugging.
+ *
+ * Here are some examples:
+ *
+ *  IPv4 netmask in struct sockaddr_in
+ * +------------------------------------+
+ * | sin_len, indicates length          |
+ * +------------------------------------+
+ * | sin_family, contains no valid data |
+ * +------------------------------------+
+ * | sin_port, contains no valid data   |
+ * |                                    |
+ * +------------------------------------+
+ * | sin_addr, contains one to four     |
+ * | bytes of netmask, indicated by     |
+ * | (sin_len - (this field's offset)). |
+ * | The remaining netmask bytes are    |
+ * | inferred to be zero.               |
+ * +------------------------------------+
+ *
+ *  IPv6 netmask in struct sockaddr_in6
+ * +------------------------------------+
+ * | sin6_len, indicates length         |
+ * +------------------------------------+
+ * | sin6_family, contains no valid data|
+ * +------------------------------------+
+ * | sin6_port, contains no valid data  |
+ * |                                    |
+ * +------------------------------------+
+ * | sin6_flowinfo, contains no valid   |
+ * | data                               |
+ * |                                    |
+ * |                                    |
+ * +------------------------------------+
+ * | sin6_addr, contains one to sixteen |
+ * | bytes of netmask, indicated by     |
+ * | (sin6_len - (this field's offset)).|
+ * | The remaining netmask bytes are    |
+ * | inferred to be zero.               |
+ * +------------------------------------+
+ */
```

Comment:
Great comments in this and the next section!  We'll have to keep an eye on this functionality as the OS changes.

Reply:
And make sure it's crappy wine before you throw it.